examples/applications/plot_cyclical_feature_engineering.py

```diff
@@ -0,0 +1,693 @@
+
+"""
+==========================
+Cyclic feature engineering
+==========================
+
+This notebook introduces different strategies to leverage time-related features
+for a bike sharing demand regression task that is highly dependent on business
+cycles (days, weeks, months) and yearly season cycles.
+
+In the process we introduce how to perform periodic feature engineering using
+the :class:`sklearn.preprocessing.SplineTransformer` class and its
+`extrapolation="periodic"` option.
+
+"""
+# %%
+# Data exploration on the Bike Sharing Demand dataset
+# ---------------------------------------------------
+#
+# We start by loading the data from the OpenML repository.
+from sklearn.datasets import fetch_openml
+
+bike_sharing = fetch_openml("Bike_Sharing_Demand", version=2, as_frame=True)
+df = bike_sharing.frame
+
+# %%
+# To get a quick understanding of the periodic patterns of the data, let us
+# have at the average weekly demand.
+#
+# Note that the week starts on a Sunday, during the week-end. We can clearly
+# distinguish the commute patterns in the morning and evenings of the work days
+# and the leisure use of the bikes on the weekends with a more spread peak
+# demand around the middle of the days:
+import matplotlib.pyplot as plt
+
+
+average_week_demand = df.groupby(["weekday", "hour"]).mean()["count"]
+average_week_demand.plot(figsize=(12, 4))
+_ = plt.title("Average number of bike rentals during the week")
+
+
+# %%
+#
+# The target of the prediction problem is the absolute count of bike rentals on
+# a hourly basis:
+df["count"].max()
+
+# %% [markdown]
+#
+# Let's rescale the target variable (absolute value of the demand) to predict a
+# relative demand so that the mean absolute error is more easily interpreted as
+# a fraction of the maximum demand and avoid numerical issues with least
+# squares regression.
+
+# %%
+y = df["count"] / 1000
```

Comment:
I used the least square loss for fitting with `RidgeCV` just because it's much faster and efficient to fit compared to a conditional median estimator or an Huber loss model.

But as a user I find it much more natural to interpret the prediction in terms of average absolute error and in particular to phrase it as a percentage of the maximum demand instead of using an absolute scale.

Reply:
Scaling `y` before fitting is probably not really needed from a numerical point of view, I agree. It just makes reporting relative errors more natural.

Reply:
I think it won't matter much in this example, but in using absolute error, you evaluate how good the model predicts the median which is not really what you have modelled.

How about a short note like "In what follows, we use the average absolute error for model comparison as an intuitive error quantity. Note that using the squared error would be more appropriate for actual model selection."

Reply:
I added a note and changed the evaluate function to also report the cross-validated RMSE in addition to the cross-valdiated MAE. Let me know what you think.

Reply:
The note is a good match. (Not-) Sorry for being pedantic about hits issue, again and again:smirk: